test/TestApps/PushNotificationsTestApp/main.cpp

```diff
@@ -0,0 +1,349 @@
+    }
+    catch (...)
+    {
+        auto registerActivatorException = hresult_error(to_hresult());
+        hr = registerActivatorException.code();
+    }
+    return hr == E_INVALIDARG;
+}
+
+bool UnregisterActivatorNullBackgroundRegistration()
+{
+    winrt::hresult hr = S_OK;
+    try
+    {
+        PushNotificationRegistrationToken badToken{ 0, nullptr };
+        PushNotificationManager::UnregisterActivator(badToken, PushNotificationRegistrationOption::PushTrigger);
+    }
+    catch (...)
+    {
+        auto registerActivatorException = hresult_error(to_hresult());
+        hr = registerActivatorException.code();
+    }
+    return hr == HRESULT_FROM_WIN32(ERROR_NOT_FOUND);
+}
+
+bool MultipleRegisterActivatorTest()
+{
+    winrt::hresult hr = S_OK;
+    try
+    {
+        PushNotificationActivationInfo info(
+            PushNotificationRegistrationOption::PushTrigger | PushNotificationRegistrationOption::ComActivator,
+            c_fakeComServerId); // same clsid as app manifest
+
+        appToken = PushNotificationManager::RegisterActivator(info);
+    }
+    catch (...)
+    {
+        auto registerActivatorException = hresult_error(to_hresult());
+        hr = registerActivatorException.code();
+    }
+    return hr == E_INVALIDARG;
+}
+
+bool runUnitTest(std::string unitTest)
+{
+    switch (switchMapping[unitTest])
+    {
+    case UnitTest::channelRequestUsingNullRemoteId:
+        return ChannelRequestUsingNullRemoteId();
+
+    case UnitTest::channelRequestUsingRemoteId:
+        return ChannelRequestUsingRemoteId();
+
+    case UnitTest::multipleChannelClose:
+        return MultipleChannelClose();
+
+    case UnitTest::multipleChannelRequestUsingSameRemoteId:
+        return MultipleChannelRequestUsingSameRemoteId();
+
+    case UnitTest::multipleChannelRequestUsingMultipleRemoteId:
+        return MultipleChannelRequestUsingMultipleRemoteId();
+
+    case UnitTest::activatorTest:
+        return ActivatorTest();
+
+    case UnitTest::registerActivatorNullDetails:
+        return RegisterActivatorNullDetails();
+
+    case UnitTest::registerActivatorNullClsid:
+        return RegisterActivatorNullClsid();
+
+    case UnitTest::unregisterActivatorNullToken:
+        return UnregisterActivatorNullToken();
+
+    case UnitTest::unregisterActivatorNullBackgroundRegistration:
+        return UnregisterActivatorNullBackgroundRegistration();
+
+    case UnitTest::multipleRegisterActivatorTest:
+        return MultipleRegisterActivatorTest();
+
+    default:
+        return false;
+    }
+}
+
+// Cleanup function for exiting test app.
+void UnregisterClsid()
+{
+    if (appToken)
+    {
+        ::CoRevokeClassObject(static_cast<DWORD>(appToken.Cookie()));
+        PushNotificationManager::UnregisterActivator(appToken, PushNotificationRegistrationOption::PushTrigger);
```

Comment:
In the scenario where we UnregisterActivator with the token, we would have to set the token to null after so we don't call UnregisterActivator twice. If the test crashes right before the null check, we would be calling UnregisterActivator a second time.

Reply:
This is in reference to ActivatorTest.

Reply:
I don't see how explicitly calling CoRevokeClassObject helps with the problem of double-unregister. This code doesn't ever set `appToken = nullptr` , so it still could unregister twice if called twice.

You could use this:

```cpp
PushNotificationManager::UnregisterActivator(std::exchange(appToken, nullptr),
    PushNotificationRegistrationOption::PushTrigger);
```

to set appToken to nullptr while extracting the previous value.

Can you explain why the code you have here is better than

```cpp
if (appToken)
{
    PushNotificationManager::UnregisterActivator(appToken,
        PushNotificationRegistrationOption::PushTrigger |
        PushNotificationRegistrationOption::ComActivator);
|
```
?

Reply:
The goal of doing it this way was to avoid hitting the failure case of ComActivator from UnregisterActivator being called twice. In this way, we would avoid the failure scenario but I think the other way is better. Will fix.

Reply:
But how did simple way cause the ComActivator to be called twice? UnregisterClsid is called only once.

Reply:
Users should call RegisterActivator at the beginning of their app and UnregisterActivator at the end of the app. For this test, we unregister the current activator to test another clsid being registered/unregistered. The issue was that I wasn't sure on how to handle cleanup of this test. I have a cleanup scope_exit function take care of remaining activators if they weren't unregistered. I knew that if UnregisterActivator was called with a token that was already revoked, the function would fail, but not throw. I just wanted to avoid a function call that would process a failure case in the cleanup function, but it is correct to unregister with the flags it was registered with.